### ISSUE TITLE
Question: how to replace tag with another tag with inner text?

### ISSUE THREAD

**Issue by IvanPizhenko** [OWASP/java-html-sanitizer#230]
I want to completely replace tag if I encounter certain conditions. I have found I have found example in the `HtmlPolicyBuilder` Javadoc (https://www.javadoc.io/doc/com.googlecode.owasp-java-html-sanitizer/owasp-java-html-sanitizer/20180219.1/org/owasp/html/HtmlPolicyBuilder.html)
```java
new HtmlPolicyBuilder()
   .allowElement(
     new ElementPolicy() {
       public String apply(String elementName, List<String> attributes){
         attributes.add("class");
         attributes.add("header-" + elementName);
         return "div";
       }
     },
     "h1", "h2", "h3", "h4", "h5", "h6")
   .build(outputChannel)
```
which shows how to use custom `ElementPolicy`. In theory, I could change list of attributes and return new tag name, similar to what is shown in the example. But in my case that’s not enough. I need to place a text inside the attribute. So for example, if I had:

```html
<img src="https://example.com/some_image.jpg"></img>
```
I want to replace this with
```html
<a href="https://example.com/some_image.jpg">some_image</a>
```
I was reading Javadoc, trying to figure out how to do it, and searching for examples on the Internet, but so far I could not find a way to place inner text. Is there a way to do it?

Same posted on [SO](https://stackoverflow.com/questions/67976114/owasp-java-html-sanitizer-replace-tag-with-inner-text)

**Comment by myin142** [comment]
You could replace it with a preprocessor, I am currently doing it like that.
```java
            .withPreprocessor(r -> new HtmlStreamEventReceiverWrapper(r) {
                @Override
                public void openTag(@NotNull String elementName, @NotNull List<String> attrs) {
                    if ("a".equals(elementName)) {
                        super.openTag("span", attrs);
                    } else {
                        super.openTag(elementName, attrs);
                    }
                }
            })

```

**Comment by IvanPizhenko** [comment]
@myin142 Thank you for you answer, but it is not a problem to replace tag, but to place inner text inside the tag. I have ended up with using Jsoup for that.

**Comment by mikesamuel** [comment]
@IvanPizhenko I was actually proposing a way to affect the innerHTML of the tag, by providing direct access to a scoped [HtmlStreamEventReceiver](https://github.com/OWASP/java-html-sanitizer/blob/ad287c39f6102d875ad55ecab3094389bb3036d4/src/main/java/org/owasp/html/HtmlStreamEventReceiver.java#L36).

That would allow specifying `innerText`/`textContent` by just calling `.text` which takes a string of text/plain.

But one could construct more complicated internals by using the open and close tag methods.

**Comment by IvanPizhenko** [comment]
@mikesamuel Is that implemented? If yes, can you please show a short working code example (like @myin142 provided above)?

**Comment by mikesamuel** [comment]
@IvanPizhenko  No.  I was thinking that I could add something that would let you do that.

The limitation that this library has, compared to JSoup, is that it operates as a streaming filter left to right.

That means it has a better memory footprint, and is less prone to denial of service, but it does mean that you can't look at a node's content when deciding what to do with it.

So I can offer some simple options to prepend/append/replace the content with something specified by a policy, but I cannot allow arbitrary rearrangement.

**Comment by IvanPizhenko** [comment]
@mikesamuel well, let's try it. Please do the changes you are talking about and provide a code example how to use them.

**Comment by WuXian-Allison** [comment]
Implement a processor as below, the inner text can be replaced.
```

    private static final PolicyFactory VALID_TAGS_POLICY = new HtmlPolicyBuilder()
            .withPreprocessor((HtmlStreamEventReceiver r) -> new HtmlStreamEventReceiverWrapper(r) {
                String newText = "";

                @Override
                public void closeTag(String elementName) {
                    // If this element is disallowed, clear it's content
                    if (!VALID_TAGS_SET.contains(elementName)) {
                        newText = "";
                    }
                    r.text(newText);
                    r.closeTag(elementName);
                }

                @Override
                public void text(String text) {
                    newText = text;
                }
            })
            .allowElements(VALID_TAGS)
            .toFactory();

```

**Comment by IvanPizhenko** [comment]
@WuXian-Allison Thank you! This is interesting idea!

**Comment by jmanico** [comment]
Backlog triage ahead of the next release.

Answered in-thread — @WuXian-Allison posted a working preprocessor that replaces both the tag and its inner text, and @myin142 posted the simpler tag-only version. The library is a streaming filter, so replacing content is done by wrapping `HtmlStreamEventReceiver` rather than by an `ElementPolicy` (which cannot see a node's content).

Thanks all. #381 tracks a related gap — exposing more of the policy building blocks publicly.
